plugins/org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/ui/labeling/LabelProviderFragment2.xtend

```diff
@@ -0,0 +1,235 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.xtext.generator.ui.labeling
+
+import com.google.inject.Inject
+import org.eclipse.xtend.lib.annotations.Accessors
+import org.eclipse.xtext.Grammar
+import org.eclipse.xtext.xtext.generator.AbstractGeneratorFragment2
+import org.eclipse.xtext.xtext.generator.CodeConfig
+import org.eclipse.xtext.xtext.generator.XtextGeneratorNaming
+import org.eclipse.xtext.xtext.generator.model.FileAccessFactory
+import org.eclipse.xtext.xtext.generator.model.GuiceModuleAccess
+import org.eclipse.xtext.xtext.generator.model.TypeReference
+import org.eclipse.xtext.xtext.generator.xbase.XbaseUsageDetector
+
+import static extension org.eclipse.xtext.GrammarUtil.*
+
+/**
+ * Contributes the Labeling provider stub, either in Xtend or Java language.
+ * 
+ * @author Christian Schneider - Initial contribution and API
+ */
+class LabelProviderFragment2 extends AbstractGeneratorFragment2 {
+
+	private static val XBASE_LABEL_PROVIDER = 
+		"org.eclipse.xtext.xbase.ui.labeling.XbaseLabelProvider"
+
+	private static val XBASE_DESCRIPTION_LABEL_PROVIDER = 
+		"org.eclipse.xtext.xbase.ui.labeling.XbaseDescriptionLabelProvider"
+
+	@Inject
+	extension XtextGeneratorNaming
+
+	@Inject
+	extension XbaseUsageDetector
+
+	@Inject
+	extension CodeConfig
+
+	@Inject
+	FileAccessFactory fileAccessFactory
+
+	@Accessors
+	private boolean generateStub = true;
+
+	def protected TypeReference getEObjectLabelProviderClass(Grammar g) {
+		return new TypeReference(
+			g.eclipsePluginBasePackage + ".labeling." + g.simpleName + "LabelProvider"
+		)
+	}
+
+	def protected TypeReference getEObjectLabelProviderSuperClass(Grammar g) {
+		if (g.inheritsXbase)
+			new TypeReference(XBASE_LABEL_PROVIDER) 
+		else defaultEObjectLabelProviderSuperClass
+	}
+
+	/**
+	 * Extra getter facilitates customization by overriding. 
+	 */
+	def protected TypeReference getDefaultEObjectLabelProviderSuperClass() {
+		new TypeReference("org.eclipse.xtext.ui.label.DefaultEObjectLabelProvider")
+	}
+
+
+	def protected TypeReference getDescriptionLabelProviderClass(Grammar g) {
+		return new TypeReference(
+			g.eclipsePluginBasePackage + ".labeling." + g.simpleName + "DescriptionLabelProvider"
+		)
+	}
+
+	def protected TypeReference getDescriptionLabelProviderSuperClass(Grammar g) {
+		if (g.inheritsXbase)
+			new TypeReference(XBASE_DESCRIPTION_LABEL_PROVIDER) 
+		else defaultDescriptionLabelProviderSuperClass
+	}
+
+	/**
+	 * Extra getter facilitates customization by overriding. 
+	 */
+	def protected TypeReference getDefaultDescriptionLabelProviderSuperClass() {
+		new TypeReference("org.eclipse.xtext.ui.label.DefaultDescriptionLabelProvider")
+	}
+
+
+	override generate() {
+		if (!generateStub && !grammar.inheritsXbase) {
+			// the binding of the DefaultEObjectLabelProvider and the 
```

Comment:
Same argument against early exit here. It would be more readable in the positive condition

```
if (generateStub || inheritsXbase) {
  bindDescriptionLabelProvider
}